### PR TITLE
[release/2.8] bump up golang version

### PR DESCRIPTION
**Pull request by wy65701436**
upgrade go version to v1.18.8

Signed-off-by: Wang Yan <wangyan@vmware.com>

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -25,21 +25,21 @@ jobs:
     - name: Set up Go
       uses: actions/setup-go@v2
       with:
-        go-version: 1.16.*
+        go-version: 1.18.*
```
Comment (thaJeztah) [inline .github/workflows/ci.yml:28]: I recall we had some discussion about this when we did the last release, but could you change the go-versions to have the exact version? (`1.18.8`); that way it's (also when doing releases) transparent "what version was used" for a release, instead of "what happened to be the version at the time".

File: Dockerfile [hunk 2]
```diff
@@ -1,49 +1,59 @@
-# syntax=docker/dockerfile:1.3
+# syntax=docker/dockerfile:1
 
-ARG GO_VERSION=1.16.15
-ARG GORELEASER_XX_VERSION=1.2.5
+ARG GO_VERSION=1.18
```
Comment (thaJeztah) [inline Dockerfile:3]: Same for this one. If anyone else wants to build from the tag and wants to use a different version of Go, they'll be able to use the build-args to override, so pinning to an exact version should not limit them.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (wy65701436) [reply .github/workflows/ci.yml]: updated.

File: Dockerfile (hunk 2 above)
Comment (wy65701436) [reply Dockerfile]: updated.

File: registry/client/transport/http_reader.go [hunk 3]
```diff
@@ -19,24 +19,23 @@ var (
 )
 
 // ReadSeekCloser combines io.ReadSeeker with io.Closer.
-type ReadSeekCloser interface {
-	io.ReadSeeker
-	io.Closer
-}
+//
+// Deprecated: use [io.ReadSeekCloser].
+type ReadSeekCloser = io.ReadSeekCloser
```
Comment (thaJeztah) [inline registry/client/transport/http_reader.go:24]: - It looks like this is backporting https://github.com/distribution/distribution/pull/3788 - was that needed to make it work with go1.18, or accidentally included in this commit? (if it's not needed for the go1.18 update, it's better to keep it in a separate PR)

Comment (thaJeztah) [reply registry/client/transport/http_reader.go]: Yes, it makes it more visible that it was back ported; and it allows you to leave a comment on the PR _why_ it needs to be backported

Comment (wy65701436) [reply registry/client/transport/http_reader.go]: I will come up with another PR to back port it, just keep this one as an upgrade to golang.

**Review by milosgajdos — APPROVED** [review]

File: registry/client/repository.go [hunk 4]
```diff
@@ -114,9 +114,7 @@ func (r *registry) Repositories(ctx context.Context, entries []string, last stri
 			return 0, err
 		}
 
-		for cnt := range ctlg.Repositories {
-			entries[cnt] = ctlg.Repositories[cnt]
-		}
+		copy(entries, ctlg.Repositories)
```
Comment (thaJeztah) [inline registry/client/repository.go:117]: This looks unrelated as well; looks like it was part of https://github.com/distribution/distribution/commit/fbdfd1ac3580f1aafdbd17c09ebd02c86ccfc4a1; which is 

- https://github.com/distribution/distribution/pull/3711

Not sure if we want to backport that as a whole (it seems rather large), but if we want, should be separate.

File: releases/v2.8.1.toml [hunk 5]
```diff
@@ -18,7 +18,7 @@ See changelog below for full list of changes.
 
 ### CI
 * ci: use proper git ref for versioning [#3595](https://github.com/distribution/distribution/pull/3595)
-* Go: make Go version explicit and pin it to the latest 1.16 release [3604](https://github.com/distribution/distribution/pull/3604)
+* Go: make Go version explicit and pin it to the latest 1.18 release [3604](https://github.com/distribution/distribution/pull/3604)
```
Comment (thaJeztah) [inline releases/v2.8.1.toml:21]: This should be reverted, as it's updating changes for the v2.8.1 release (which did use go1.16)

File: version/version.sh [hunk 6]
```diff
@@ -17,7 +17,7 @@ var Package = "$(go list)"
 // Version indicates which version of the binary is running. This is set to
 // the latest release tag by hand, always suffixed by "+unknown". During
 // build, it will be replaced by the actual version. The value here will be
-// used if the registry is run after a go get based install.
+// used if the registry is run after a go install based install.
```
Comment (thaJeztah) [inline version/version.sh:20]: Looks like we didn't update this in the `main` branch; we should do it there as well;
https://github.com/distribution/distribution/blob/92d136e113cf34fcfd5f0c21fadf6eac2f39e803/version/version.sh#LL20

File: script/validate/dco [hunk 7]
```diff
@@ -4,7 +4,7 @@ set -eu -o pipefail
 
 if ! command -v git-validation; then
     >&2 echo "ERROR: git-validation not found. Install with:"
-    >&2 echo "    go get -u github.com/vbatts/git-validation"
+    >&2 echo "    go install github.com/vbatts/git-validation"
```
Comment (thaJeztah) [inline script/validate/dco:7]: Should have either `@latest` added, or (as it's only an information message) we can decide to revert it to keep the diff small for this branch.

File: BUILDING.md [hunk 8]
```diff
@@ -26,7 +26,7 @@ development environment setup. Please follow [How to Write Go Code](https://gola
 for proper setup. If done correctly, you should have a GOROOT and GOPATH set in the
 environment.
 
-If a Go development environment is setup, one can use `go get` to install the
+If a Go development environment is setup, one can use `go install` to install the
```
Comment (thaJeztah) [inline BUILDING.md:29]: The instructions themselves below haven't been updated, so I suggest reverting the changes in this file

But we should update the one in `main` to be updated for `go modules` (it still outlines using `GOPATH` etc).

File: BUILDING.md [hunk 9]
```diff
@@ -71,7 +71,7 @@ commands, such as `go test`, should work per package (please see
 A `Makefile` has been provided as a convenience to support repeatable builds.
 Please install the following into `GOPATH` for it to work:
 
-    go get github.com/golang/lint/golint
+    go install github.com/golang/lint/golint
```
Comment (thaJeztah) [inline BUILDING.md:74]: Same as above for these; also I don't think we still use `golint`, so this is all rather obsolete, but as this branch is mostly in "maintenance", I don't think we should spend much time on rewriting everything, so would suggest reverting the changes and just let them be.

File: context/http.go [hunk 10]
```diff
@@ -246,11 +246,7 @@ func (ctx *muxVarsContext) Value(key interface{}) interface{} {
 			return ctx.vars
 		}
 
-		if strings.HasPrefix(keyStr, "vars.") {
-			keyStr = strings.TrimPrefix(keyStr, "vars.")
-		}
-
-		if v, ok := ctx.vars[keyStr]; ok {
+		if v, ok := ctx.vars[strings.TrimPrefix(keyStr, "vars.")]; ok {
```
Comment (thaJeztah) [inline context/http.go:249]: This change is unrelated, and looks to be part of https://github.com/distribution/distribution/commit/3b391d32908f3523b525ac9e5785a24a9f488919;

- https://github.com/distribution/distribution/pull/3769

I suggest reverting this; if we want the other changes, we should backport it as a whole and having this partial change will complicate that (result in merge conflicts)

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks! Looks like there's still some unrelated changes and changes not in `main` in here; I left comments inline

File: registry/storage/driver/inmemory/mfs.go [hunk 11]
```diff
@@ -279,6 +279,9 @@ func (f *file) sectionReader(offset int64) io.Reader {
 }
 
 func (f *file) ReadAt(p []byte, offset int64) (n int, err error) {
+	if offset >= int64(len(f.data)) {
+		return 0, io.EOF
+	}
```
[the remark this answers is not in the document]
Comment (wy65701436) [reply registry/storage/driver/inmemory/mfs.go:284]: this should be in, otherwise one ut will panic.

File: registry/client/repository.go (hunk 4 above)
Comment (wy65701436) [reply registry/client/repository.go:117]: copy(entries, ctlg.Repositories) is for goci-lint failure.

File: version/version.sh (hunk 6 above)
Comment (wy65701436) [reply version/version.sh:20]: will raise pr on main later.

File: script/validate/dco (hunk 7 above)
Comment (wy65701436) [reply script/validate/dco:7]: sure, it's ok

File: context/http.go (hunk 10 above)
Comment (wy65701436) [reply context/http.go:249]: again, it's to fix a goci-lint failure.

File: releases/v2.8.1.toml (hunk 5 above)
Comment (wy65701436) [reply releases/v2.8.1.toml:21]: reverted.

File: BUILDING.md (hunk 8 above)
Comment (wy65701436) [reply BUILDING.md]: reverted.

File: BUILDING.md (hunk 9 above)
Comment (wy65701436) [reply BUILDING.md:74]: reverted.

File: registry/storage/driver/inmemory/mfs.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (wy65701436) [reply registry/storage/driver/inmemory/mfs.go:284]: This is not directly related to Go 1.18, but without this change, UT would have failed.

File: registry/client/repository.go (hunk 4 above)
Comment (wy65701436) [reply registry/client/repository.go:117]: It's not easy to decouple the fix for golangci-lint from the version upgrade of go, because there is some correlation between the versions of go and golangci-lint.

So, in general, you would want us to fix all the prerequisites for the go upgrade, and then have this pr focus only on the go version upgrade.

File: registry/storage/driver/inmemory/mfs.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (milosgajdos) [reply registry/storage/driver/inmemory/mfs.go:284]: @wy65701436 do you want to rebase now that we've merged https://github.com/distribution/distribution/pull/3815

**Comment by thaJeztah** [comment]
I had a look at the changes included in this PR, and where they came from; opened an alternative with those changes applied / cherry-picked separately (separate commits), and rebased this PR on top to check for the diff; see https://github.com/distribution/distribution/pull/3903
